csvcubed/csvcubed/constraints/virtualiser.py

```diff
@@ -0,0 +1,196 @@
+        """
+        Run a sparql query against the simple metadata graph
+        """
+        try:
+            result: SPARQLResult = self.metadata_graph.query(query_str)
+        except ParseException as err:
+            raise ParseException(f"Failed to parse query {query_str}") from err
+
+        return result
+
+    def _subset_index_yieder(self) -> Tuple[int, int]:
+        """
+        Given a csv of unknown length of rows, yields appropriate
+        start:end rows for reading a dataframe or series taken from that
+        csv as an appropriately performant slice of rows.
+
+        i.e don't allow reading in a 3 million rows at once, we'll
+        regret that.
+        """
+
+        # only calculate this once per csv
+        if not self.row_count:
+            self.row_count = utils.get_csv_length(self.csv_path)
+
+        # only calculate this once per csv
+        if not self.start_end_indicies:
+            self.start_end_indicies = []
+            i = 1
+            last_row = 0
+
+            while True:
+                if i * MAXIMUM_ROW_SUBSET > self.row_count:
+                    start_end_index = (last_row, self.row_count + 1)
+                    self.start_end_indicies.append(start_end_index)
+                    break
+                else:
+                    start_end_index = (last_row, i * MAXIMUM_ROW_SUBSET)
+                    self.start_end_indicies.append(start_end_index)
+                    last_row = i * MAXIMUM_ROW_SUBSET
+                    i += 1
+
+        # TODO: probably unnessary/OTT, needs some thought.
+        # copy as a precuation ony, I'm concerned about mutabilty
+        # if we're sourcing multiple genrators against the same list,
+        # remove if/once it's been given some thought.
+        subset_yielder = copy.deepcopy(self.start_end_indicies)
+        for start_end_index in subset_yielder:
+            yield start_end_index[0], start_end_index[1]
+
+    def _column_subset_yielder(self, column_label: str) -> pd.DataFrame:
+        """
+        Generator that returns a sinele column dataframe representing a
+        finite subset of the complete dataframe equal to a maximum
+        rowcount of MAXIMUM_ROW_SUBSET
+        """
+
+        first_iteration = True
+
+        for start, end in self._subset_index_yieder():
+            # Keep track of the headers from the first iteraton and reuse them
+            # (names) for later dataframe slices that don't have a header row
+            if first_iteration:
+                one_col_df: pd.DataFrame = pd.read_csv(
+                    self.csv_path, usecols=[column_label], skiprows=start, nrows=end
+                )
+                headers = one_col_df.columns.values
+                first_iteration = False
+            else:
+                one_col_df: pd.DataFrame = pd.read_csv(
+                    self.csv_path,
+                    names=headers,
+                    usecols=[column_label],
```

Comment:
Ah, so this only reads one column at a time from the CSV?

Reply:
yeah that was the idea I was playing with, only ever be reading in x rows of a given column at a time as a means to minimise/control memory footprint. I was worried the overhead of just x rows would fluxuate based on number of columns.